### PR TITLE
Rewriter API for NPM (rebased)

### PR DESCRIPTION
**Pull request by beta-ziliani**
(Sending it again with the new master, so there's no noise about the API.)

I added two examples to the user guides to check their usage, but still the documentation in general is lacking, or the code ugly. The idea is to have a quick look and see if this is the kind of API we want to have for these scenarios.

It's expeted to fail in CI for lack of documentation.

### CODE REVIEW

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 1]
```diff
@@ -0,0 +1,96 @@
+{%- if rendering_in_stubs -%}
+  export abstract class BaseRewriter { }
+{%- else -%} 
+  import { NonterminalKind, NonterminalNode, Node, TerminalNode, TerminalKind, Edge } from "../index.mjs";
+
+  export abstract class BaseRewriter {
+    
+    public rewriteNode(node: Node): Node | undefined {
+      if (node instanceof NonterminalNode) {
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: nit: suggest using `NodeType` for exhaustiveness, in case we change the implementation/types later:

```ts
switch (node.type) {
  case NodeType.TerminalNode:
    return this.rewriteTerminalNode(node);
  case NodeType.NonterminalNode:
    return this.rewriteNonterminalNode(node);
}
```

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    public rewriteTerminalNode(node: TerminalNode): Node | undefined {
+      switch (node.kind) {
+        {% for terminal in model.kinds.terminal_kinds %}
+        case TerminalKind.{{ terminal.id }}:
+          return this.rewrite{{- terminal.id -}}(node);
+        {% endfor %}
+        case TerminalKind.Unrecognized:
+          return this.rewriteUnrecognized(node);
+        case TerminalKind.Unrecognized:
+          return this.rewriteMissing(node);
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: is this `TerminalKind.Missing`?

Is this the reason the `switch` is not exhaustive? can we then remove the `throw Error()` below?

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      return node;
+    }
+  
+    public rewriteMissing(node: TerminalNode): Node | undefined {
+      return node;
+    }
+  
+    protected rewriteChildren(kind: NonterminalKind, node: NonterminalNode): NonterminalNode {
+      const newChildren: Array<Edge> = new Array<Edge>();
+      var anyChildChanged = false;
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: suggest using `let` everywhere instead of `var`, as the latter uses deprecated (function) scoping rules.

Comment (OmarTawfik) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: side note: maybe we should add `eslint` as a new linter to the codebase? 🤔
It should help fix these things automatically as we are now starting to have more TypeScript in the codebase...

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 (hunk 1 above)
Comment (OmarTawfik) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: Same suggestion below..

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    public rewriteUnrecognized(node: TerminalNode): Node | undefined {
+      return node;
+    }
+  
+    public rewriteMissing(node: TerminalNode): Node | undefined {
+      return node;
+    }
+  
+    protected rewriteChildren(kind: NonterminalKind, node: NonterminalNode): NonterminalNode {
+      const newChildren: Array<Edge> = new Array<Edge>();
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: TOL: possible perf optimization?

This algorithm creates and populates `newChildren` in all cases. Since most nodes will be the same, is it better to create it only when the first different node is found (cloning the previous children so far)?

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  public override rewriteIdentifier(terminalNode: TerminalNode): Node | undefined {
+    if (!this.insideContract) {
+      return TerminalNode.create(terminalNode.kind, terminalNode.unparse() + "_new");
+    } else {
+      return terminalNode;
+    }
+  }
+
+  public override rewriteContractDefinition(node: NonterminalNode): Node | undefined {
+    this.insideContract = true;
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: I suggst adding a test for when recursive visitors depend on each other. For example, renaming identifiers within a contract...

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    const newChildren = node.children().map((edge) => {
+      if (edge.label == EdgeLabel.Name) {
+        return Edge.createWithTerminal(EdgeLabel.Name, TerminalNode.create(TerminalKind.Identifier, "NewName"));
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: IIUC, `rewriteContractDefinition()` stops the recursion at `ContractDefinition`, without calling `rewriteIdentifier()`.. Are they related? If not, I suggest separating each rewriter with its tests..

```ts
describe("Foo Rewriter", () => {
  class FooRewriter extends BaseRewriter { }

  it("....", () => { });
  
  it("....", () => { });
})

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left one concern on the exhaustive of the visitor (missing `MISSING`). But otherwise, implementation looks great!

Didn't review the docs yet per our conversation..

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 (hunk 2 above)
Comment (beta-ziliani) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: 🤦

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 5 above)
Comment (beta-ziliani) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: Solved in af90ac40fd85edd48073c52d3264ebdeec3a61e9.

That said, I realize that we don't have the context from where the node comes from. Maybe we can extend each `rewrite` method with an `EdgeLabel`?

So, for instance, in the example in that commit the rewrite of an identifier is turned on when processing a contract, and then after the first rewrite (the name), it turns it off. But this only works because the name is the first identifier found in a contract. Ideally, we should check we're in the `name:` label.

OTOH the problem with this idea is that it still misses context when traversing further down. For instance, function definitions also have a `name:` label. To really solve it, we need something more sophisticated... Or leave it to the user.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 6 above)
Comment (beta-ziliani) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: I went simpler: each test has its own rewriter (because there's no real reuse). (af90ac40fd85edd48073c52d3264ebdeec3a61e9)

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 (hunk 3 above)
Comment (beta-ziliani) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: I tried to add eslint https://github.com/manastech/slang/commit/8fdd78d4f078d7bebe41b8518a6150ee87b2c009
but failed to make it work with jest tests. Tried so many things, but none worked, so I'm just giving up for now.

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      children.forEach((child, index) => {
+        const newChild = this.rewriteNode(child.node);
+        if (newChild == undefined) {
+          // node was removed, mark the removal
+          newChildren = newChildren || new Map<number, Edge | "delete">();
+          newChildren.set(index, "delete");
+        } else {
+          let edge;
+          if (newChild.id == child.node.id) {
+            edge = child;
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: not sure I understand this bit:
Shouldn't we be `return`ing if the ID didn't change? this would proceed to create and populate `newChildren` anyways.

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                break;
+            }
+          }
+          newChildren = newChildren || new Map<number, Edge | "delete">();
+          newChildren.set(index, edge);
+        }
+      });
+
+      if (newChildren != undefined) {
+        let deleted = 0;
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: I wonder why the second iteration and the `delete` literal here?
If a node was deleted, we can simply initialize `newChildren` with the identical children so far (before current `index`) and skip the current one.
For subsequent nodes, `newChildren` will be initialized already, so the original loop would know to append to it anyways.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 5 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: > the context from where the node comes from

The edge is a property of the parent, not the child. Therefore, I suggest that rewriters check the edge to do so in the parent visit directly.
So, for accurately rewriting a contract's name, we should be doing so in `rewriteContractDefinition()` and checking/replacing the specific child with `EdgeLabel.Name`.
Same thing with `rewriteFunctionDefinition()`, etc...

WDYT?

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    public override rewriteIdentifier(terminalNode: TerminalNode): Node | undefined {
+      return TerminalNode.create(terminalNode.kind, terminalNode.unparse() + "New");
+    }
+  }
+
+  const terminalNode = TerminalNode.create(TerminalKind.Identifier, "test");
+  const rewriter = new IdRewriter();
+  const result = rewriter.rewriteNode(terminalNode);
+  assert.ok(result);
+  if (result instanceof TerminalNode) {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: For this + others below: we have existing utilities to do that in the public API:

```ts
assertTerminalNode(result, TerminalKind.Identifier, "test");
```


```ts
assertNonterminalNode(result, NonterminalKind.ContractDefinition, "contract NewName {}");
```

File: documentation/public/user-guide/08-examples/05-remove-unused-definitions/examples/cleaner.mts [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import { Definition } from "@nomicfoundation/slang/bindings";
+import { BaseRewriter, Node, NonterminalNode } from "@nomicfoundation/slang/cst";
+
+export class Cleaner extends BaseRewriter {
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/05-remove-unused-definitions/examples/cleaner.mts]: I suggest renaming the class to `RemoveUnusedDefs` directly for clarity.

File: documentation/public/user-guide/08-examples/05-remove-unused-definitions/examples/cleaner.mts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  constructor(private readonly unusedDefinitions: Definition[]) {
+    super();
+  }
+
+  removeUnused(node: NonterminalNode): Node | undefined {
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/05-remove-unused-definitions/examples/cleaner.mts]: nit: Default visibility in TS is `public`:

```ts
private removeUnused(node: NonterminalNode): Node | undefined {
```

File: documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    this.functionName = undefined;
+    return recurse;
+  }
+
+  // once in the statements of a function, inject a call to the `log` function.
+  public override rewriteStatements(node: NonterminalNode): Node | undefined {
+    if (this.functionName) {
+      this.injected = true;
+
+      // collect the trivia so we respect the same indentation.
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts]: > so we respect the same indentation

I suggest skipping the trivia collection to simplify the example.
Maybe we can inject `log()` with four leading spaces directly? It would probably cut down this example to half.

In a real-world scenario, formatting/indentation would be handled by `prettier`, as it would run on the output automatically before committing the new file.

File: documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      const trivia = this.collectLeadingTrivia(node);
+      const unparsedTrivia = trivia.reduce((soFar, edge) => soFar + edge.node.unparse(), "");
+
+      // the injected code
+      const toInject = this.parser.parseNonterminal(
+        NonterminalKind.ExpressionStatement,
+        `${unparsedTrivia}log("${this.functionName}");\n`,
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts]: nit: I suggest also using a leading `\n` to prevent it with merging with a previous line comment.

File: crates/codegen/runtime/npm/package/src/runtime/cst/index.mts [hunk 14]
```diff
@@ -2,6 +2,7 @@ import * as wasm from "../../../wasm/index.mjs";
 
 export * from "./assertions.mjs";
 export * from "./extensions.mjs";
+export * from "./generated/rewriter.mjs";
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/src/runtime/cst/index.mts:5]: Since #1347 is now merged, WDYT of including the Rust rewriter in this PR?

File: crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2 (hunk 7 above)
Comment (beta-ziliani) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: Not sure I understand your question. This is run in loop for each children, so `return`ing doesn't make sense in this context. What this does is re-create the edge if the child has changed, or leave it as-is if it hasn't.

Comment (OmarTawfik) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: You are right. I missed the original intent.

But in this case, we proceed below to initialize `newChildren` and append to it regardless:

```ts
          newChildren = newChildren || new Map<number, Edge | "delete">();
          newChildren.set(index, edge);
```

If the ID is the same, should we skip initializing it, and only append to it if it is already initialized?

Comment (OmarTawfik) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: Also, should we be using `push()` instead of `set()`, in case there are deleted nodes before it?

```ts
if (weHaveANewNode) {
    newChildren = newChildren || new Map<number, Edge | "delete">();
}
if (newChildren) {
    newChildren.push(edge);
}
```

Comment (beta-ziliani) [reply crates/codegen/runtime/npm/package/src/runtime/cst/rewriter.mts.jinja2]: I changed the logic entirely, and only re-create the array if some node changed.

File: documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts (hunk 13 above)
Comment (beta-ziliani) [reply documentation/public/user-guide/08-examples/06-inject-logging/examples/logging-rewriter.mts]: On a second thought, given that the trivia is now fixed, I think adding a newline in this particular example will make the expected code look ugly.

File: crates/solidity/outputs/cargo/tests/Cargo.toml [hunk 15]
```diff
@@ -22,6 +22,7 @@ slang_solidity = { workspace = true, features = [
   "__private_ariadne_errors",
   "__private_testing_utils",
   "__private_backend_api",
+  "__private_wasm_apis",
```
Comment (beta-ziliani) [inline crates/solidity/outputs/cargo/tests/Cargo.toml:25]: Check: this shouldn't be necessary, probably a merge that went wrong.

Comment (beta-ziliani) [inline crates/solidity/outputs/cargo/tests/Cargo.toml:25]: This is a fix unrelated to this PR.

When running `infra test cargo`, it uses all features; but when running the specific package with `cargo test` it fails because this feature is required (for other tests unrelated to this PR). It's better to be explicit of the features needed by the tests.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+});
+
+test("Rewrite NonterminalNode Deep", () => {
+  class BasicRewriter extends BaseRewriter {
+    insideContract = false;
+
+    public override rewriteIdentifier(terminalNode: TerminalNode): Node | undefined {
+      if (this.insideContract) {
+        const newNode = TerminalNode.create(terminalNode.kind, terminalNode.unparse() + "New");
+        this.insideContract = false;
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: I suggest modifying `insideContract` only inside `rewriteContractDefinition()`, because then it would be `false` for other identifiers inside this contract, which is incorrect.
The bug results in `aFun` not getting renamed to `aFunNew` as expected.

```ts
    public override rewriteContractDefinition(node: NonterminalNode): Node | undefined {
      this.insideContract = true;
      const newNode = this.rewriteChildren(node);
      this.insideContract = false;
      return newNode;
    }
```

Additionally, to truly test that rewrites don't happen outside the contract, I suggest adding something else (a library?) next to it, that is not expected to be rewritten:

```solidity
contract AContract {
  function aFun() public {}
}
library ALibrary {}
```

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  const node = parse(NonterminalKind.SourceUnit, "contract AContract {\n  function aFun() public {}\n}");
+  const rewriter = new BasicRewriter();
+  const result = rewriter.rewriteNode(node);
+  assertNonterminalNode(result, NonterminalKind.SourceUnit, "contract AContractNew {\n  function aFun() public {}\n}");
+});
+
+test("Remove NonterminalNode", () => {
+  class RemovalRewriter extends BaseRewriter {
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts:98]: WDYT of another test that does the opposite? adding a function to an empty contract?

File: documentation/public/user-guide/08-examples/06-inject-logging/index.md [hunk 18]
```diff
@@ -0,0 +1,19 @@
+# 8.5. Inject logging
+
+In this example, we use the `BaseRewriter` class to achieve something different: instead of removing nodes, we edit them to insert snippets of code. In particular, we insert in every function a call to a fictitious `log` function, using the name of the function to make it a bit more interesting.
+
+In essence, the `LoggingRewriter` class below works as follow:
+
+1. When traversing a function definition, we collect the name and store it in a local variable. Then, we proceed to recurse, using the `rewriteChildren` function of `BaseRewriter`. Before returning, the name is cleared off.
+2. When traversing the list of statements of a function, we prepare the code to inject (if the name is set), by parsing the new statement: the function call to the `log` function. In order to make it properly, we _prepend_ the required trivia, to make sure the spacing is respected (although note: it will also copy newlines and comments!). We then construct the new statements of the function by prepending (note the `unshift`) the injected code into the children of the function statements.
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/06-inject-logging/index.md]: Since trivia handling is now removed, we should also remove the related text from the docs?

> 2. When traversing the list of statements of a function, we prepare the code to inject (if the name is set), by creating the new statement: the function call to the `log` function.
> 3. We then construct the new statements of the function by prepending (note the `unshift`) the injected code into the children of the function statements.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  `;
+  const node = parse(NonterminalKind.ContractDefinition, contract);
+  const rewriter = new RemovalRewriter();
+  const result = rewriter.rewriteNode(node);
+  assertNonterminalNode(result, NonterminalKind.ContractDefinition, expected);
+});
+
+function parse(kind: NonterminalKind, input: string): Node {
+  const parser = Parser.create(LanguageFacts.latestVersion());
```
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts:161]: I suggest adding a test where the rewriter doesn't change any nodes, and asserting that the returned root node has the same ID (unchanged).

File: crates/codegen/runtime/cargo/crate/src/runtime/cst/rewriter.rs.jinja2 [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+      /// Rewrites all the children of a given non-terminal node.
+      fn rewrite_children(&mut self, node: &Rc<NonterminalNode>) -> Node {
+          let mut new_children: Option<Vec<Edge>> = None;
+
+          for (index, child) in node.children.iter().enumerate() {
+              if let Some(new_child) = self.rewrite_node(&child.node) {
+                  if new_child.id() == child.node.id() {
+                      if let Some(ref mut nc) = new_children {
+                          nc.push(child.clone());
+                      }
```
Comment (OmarTawfik) [inline crates/codegen/runtime/cargo/crate/src/runtime/cst/rewriter.rs.jinja2]: why the additional clone? can we just use `new_child`?

```rust
if let Some(ref mut new_children) = new_children {
    new_children.push(new_child);
}
```

File: documentation/public/user-guide/08-examples/index.md [hunk 21]
```diff
@@ -4,6 +4,8 @@
 - [8.2. Find usages](./02-find-usages/index.md)
 - [8.3. Jump to definition](./03-jump-to-definition/index.md)
 - [8.4. Find unused definitions](./04-find-unused-definitions/index.md)
+- [8.5. Remove unused definitions](./05-remove-unused-definitions/index.md)
```
Comment (OmarTawfik) [inline documentation/public/user-guide/08-examples/index.md:7]: I also suggest adding a changeset to explain the API and how it can be used.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 16 above)
Comment (beta-ziliani) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: Note that I'm not rewriting every identifier inside the contract, just the name (the first one). If I implement your code, it will rewrite also the name of `aFun`, because the contract members are part of the children of the contract definition.

File: crates/codegen/runtime/cargo/crate/src/runtime/cst/rewriter.rs.jinja2 (hunk 20 above)
Comment (beta-ziliani) [reply crates/codegen/runtime/cargo/crate/src/runtime/cst/rewriter.rs.jinja2]: Names are unfortunate here: `new_child` is a node, while `new_children` are edges. But perhaps it's better just to create a new edge, and transfer the ownership of `new_child` to it. That's what I'm implementing now. (I also removed an unnecessary clone in the next if branch).

**Comment by beta-ziliani** [comment]
Ready for a second round @OmarTawfik . Thanks for the input!

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 16 above)
Comment (OmarTawfik) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: Wouldn't rewriting a specific terminal in a specific parent be done using labels?
Recursive checks like this are typically used when you need to rewrite a specific terminal anywhere in a tree/subtreee.

File: .changeset/green-poets-bow.md [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+class MyRewriter extends BaseRewriter {
+    public override rewriteContractDefinition(node: NonterminalNode): Node | undefined {
+        // Code to either replace (return a new node), remove (return undefined),
+        // or edit a node (return a modified copy of the given `node`)
+    }
+}
+
+const rewriter = new MyRewriter();
+const newNode = rewriter.rewriteNode(someNode);
+```
```
Comment (OmarTawfik) [inline .changeset/green-poets-bow.md]: I suggest adding a link to the appropriate user guide, to keep the changeset short:

```md
Add a rewriter API, allowing the transformation of CSTs by extending the `BaseRewriter` type, overriding the appropriate methods ([User Guide](https://nomicfoundation.github.io/slang/1.3.0/user-guide/08-examples/06-inject-logging/)).
```

**Review by OmarTawfik — APPROVED** [review]
Left one small suggestion on changests. But LGTM otherwise!

File: .changeset/green-poets-bow.md (hunk 22 above)
Comment (beta-ziliani) [reply .changeset/green-poets-bow.md]: ah I misunderstood your previous comment. Yes, this makes more sense.

File: crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts (hunk 16 above)
Comment (beta-ziliani) [reply crates/solidity/outputs/npm/tests/src/cst/rewriter.test.mts]: Indeed. But I wasn't worrying too much about that because it was testing the functionality, not the usability. Now I found a better example.
